File: bmp.m
clear all;clc;

%读取文件句柄，图片保存在当前文件夹
img = fopen('派蒙.bmp','rb');

%根据bmp图像结构读出文件头和信息头
type = fread(img,2,'char')
imgsize = fread(img,1,'long')
bfReserved1and2 = fread(img,1,'long')
bfOffBits = fread(img,1,'long')
biSize = fread(img,1,'long')
biWidth = fread(img,1,'long')
biHeight = fread(img,1,'long')
biPlanes = fread(img,1,'short')
biBitCount = fread(img,1,'short')
biCompression = fread(img,1,'long')
biSizeImage = fread(img,1,'long')
biXPelsPerMeter = fread(img,1,'long')
biYPelsPerMeter = fread(img,1,'long')
biClrUsed = fread(img,1,'long')
biClrImportant = fread(img,1,'long')
fclose(img);

%生成包含图像RGB的矩阵
img = fopen('派蒙.bmp', 'rb'); 
[data, count] = fread(img);  

%取出在data中的RGB信息，然后重新排列矩阵  
% img_data = zeros(biWidth, biHeight);  
img_data = reshape(data(bfOffBits+1:end), biWidth*3,biHeight);  
img_data = flipud(img_data'); %先转置，然后上下交换    

%分别取出RGB各个分量信息，再合成
cloB = img_data(:,1:3:end);
cloG = img_data(:,2:3:end);
cloR = img_data(:,3:3:end);
RGB(:,:,1) = cloR;
RGB(:,:,2) = cloG;
RGB(:,:,3) = cloB;
RGB = uint8(RGB);   %转换成unit格式
imshow(RGB);        %显示图片

fclose(img);

%生成RGB的txt文件，查看文件时txt需要自动换行
fid = fopen('rgb.txt','w');
for k = 1:biHeight
   for i = 1:biWidth
     fprintf(fid,'%d,%d,%d\t',img(k,3*i-2),img(k,3*i-1),img(k,3*i));
   end
     fprintf(fid,'\n');
end
fclose(fid);

File: bmp.cpp
#include <stdio.h>
#include <string>
#include <unistd.h>

using namespace std;

string message[15] = {"type", "imgsize", "bfReserved1&2", "bfOffBits", "biSize", "biWidth", "biHeight", "biPlanes", "biBitCount", "biCompression", "biSizeImage", "biXPelsPerMeter", "biYPelsPerMeter", "biClrUsed", "biClrImportant"};

void bmp_analysis(string filename, int *result)
{
    if (access(filename.c_str(), F_OK) == -1)
    {
        printf("%s not exist!\n", filename.c_str());
        return;
    }

    FILE *bmp = 0;
    if ((bmp = fopen(filename.c_str(), "rb")) == 0)
    {
        printf("Something wrong when read %s!\n", filename.c_str());
        return;
    }

    fread(&result[0], 1, sizeof(char) * 2, bmp);
    fread(&result[1], 1, sizeof(int), bmp);
    fread(&result[2], 1, sizeof(int), bmp);
    fread(&result[3], 1, sizeof(int), bmp);
    fread(&result[4], 1, sizeof(int), bmp);
    fread(&result[5], 1, sizeof(int), bmp);
    fread(&result[6], 1, sizeof(int), bmp);
    fread(&result[7], 1, sizeof(short), bmp);
    fread(&result[8], 1, sizeof(short), bmp);
    fread(&result[9], 1, sizeof(int), bmp);
    fread(&result[10], 1, sizeof(int), bmp);
    fread(&result[11], 1, sizeof(int), bmp);
    fread(&result[12], 1, sizeof(int), bmp);
    fread(&result[13], 1, sizeof(int), bmp);
    fread(&result[14], 1, sizeof(int), bmp);

    fclose(bmp);
}

int main()
{
    string filename;
    filename = "派蒙.bmp";
    int result[15] = {0};
    bmp_analysis(filename, result);
    for (int i = 0; i < 15; i++)
    {
        if (i == 0)
        {
            printf("%s: %x\n", message[i].c_str(), result[i]);
            continue;
        }
        printf("%s: %d\n", message[i].c_str(), result[i]);
    }

    return 0;
}